pkg/auth/cell.go

```diff
@@ -97,9 +124,43 @@ func newManager(params authManagerParams) error {
 		},
 	})
 
+	mapGC := newAuthMapGC(mapCache, params.IPCache)
+
+	jobGroup := params.JobRegistry.NewGroup(
+		job.WithLogger(params.Logger),
+		job.WithPprofLabels(pprof.Labels("cell", "auth")),
+	)
+
+	registerGCJobs(jobGroup, mapGC, params)
+
+	params.Lifecycle.Append(jobGroup)
+
 	return nil
 }
 
+func registerGCJobs(jobGroup job.Group, mapGC *authMapGarbageCollector, params authManagerParams) {
+	// Add identities based auth gc if k8s client is enabled
+	if params.CiliumIdentities != nil {
+		jobGroup.Add(job.Observer("auth identities gc",
+			mapGC.HandleCiliumIdentityEvent,
+			stream.FromChannel(params.CiliumIdentities.Events(context.Background())),
+		))
+	}
+
+	// Add node based auth gc if k8s client is enabled
+	if params.CiliumNodes != nil {
+		jobGroup.Add(job.Observer("auth nodes gc",
+			mapGC.HandleCiliumNodeEvent,
+			stream.FromChannel(params.CiliumNodes.Events(context.Background())),
```

Comment:
I would prefer if this would use a context that is cancelled when this stops. 

Looking at this code I think it'd be better if `Resource[T]` would also be an `Observable`. It used to be, but then I decided to simplify it as in pretty much all the cases we didn't care about completion. Implementation is simple though:

```
diff --git a/pkg/k8s/resource/resource.go b/pkg/k8s/resource/resource.go
index 8f0c75919f..b11f0a9982 100644
--- a/pkg/k8s/resource/resource.go
+++ b/pkg/k8s/resource/resource.go
@@ -17,6 +17,7 @@ import (
        "github.com/cilium/cilium/pkg/hive"
        "github.com/cilium/cilium/pkg/lock"
        "github.com/cilium/cilium/pkg/promise"
+       "github.com/cilium/cilium/pkg/stream"
 )
 
 // Resource provides access to a Kubernetes resource through either
@@ -40,6 +41,8 @@ import (
 // The resource is lazy, e.g. it will not start the informer until a call
 // has been made to Events() or Store().
 type Resource[T k8sRuntime.Object] interface {
+       stream.Observable[Event[T]]
+
        // Events returns a channel of events. Each event must be marked as handled
        // with a call to Done(), otherwise no new events for this key will be emitted.
        //
@@ -295,6 +298,10 @@ func WithErrorHandler(h ErrorHandler) EventsOpt {
        }
 }
 
+func (r *resource[T]) Observe(ctx context.Context, next func(Event[T]), complete func(error)) {
+       stream.FromChannel(r.Events(ctx)).Observe(ctx, next, complete)
+}
+
 // Events subscribes the caller to resource events.
 //
 // Each subscriber has their own queues and can process events at their own
```

Up for doing this, or shall we punt it to a follow-up PR?

Reply:
@joamaki will open a follow-up PR. i think making the k8s Resource an observable is the only logical step - otherwise i'm back to using a dedicated lifecycle a lifecycle start&stop hook...

Reply:
@joamaki i opened https://github.com/cilium/cilium/pull/25934 for this